lib/Sema/OverloadChoice.h

```diff
@@ -46,6 +46,8 @@ enum class OverloadChoiceKind : int {
   BaseType,
   /// \brief The overload choice selects a key path subscripting operation.
   KeyPathApplication,
+  /// \brief The member is looked up through DynamicMemberLookupProtocol.
+  DynamicMemberLookup,
```

Comment:
Given that this is a separate `OverloadChoiceKind`, it doesn't seem like it's going to interact at all well with `AnyObject` lookup or key paths.

For example it seems like:
```
  var o: AnyObject = ...
  _ = o[dynamicMemberLookup: "property"]
```
will work, if the subscript is marked `@objc`.

What happens if someone attempts `AnyObject` lookup on the property name, though?
```
  _ = o.property
```
Since this feature is about non-ObjC interop, perhaps `@dynamicMemberLookup` attribute checking should only allow subscripts that are not marked `@objc`?

And what happens with key paths?
```
@dynamicMemberLookup
class C {
  subscript(dynamicMember member: String) -> Int { return 7 }
}

_ = \C.testLookup
_ = \C.[dynamicMember: "hi"]
```

Reply:
Thank you for the review feedback!

This feature doesn't interact with AnyObject - AnyObject doesn't have the attribute on it, and it doesn't make sense for it to.  Also, this proposal doesn't "take away" direct use of the subscript - there are perfectly reasonable reasons to use it directly as well as through DML.

Reply:
> This feature doesn't interact with AnyObject - AnyObject doesn't have the attribute on it, and it doesn't make sense for it to.

I think some users might look at this from a different direction, especially if they know nothing about implementation details but just view `AnyObject` lookup as magic that can dynamically find things in classes.

```
@dynamicMemberLookup
@objc
class C : NSObject {
  @objc var i: Int
  @objc subscript(dynamicMember member: String) -> Int { ... magic happens ... }
}

var c = C()

c.i      // => c.i
c[dynamicMember: "x"]  // works as expected
c.x     // => c[dynamicMember: "x"]

var o: AnyObject = c
o.i      // => c.i
o[dynamicMember: "x"] // works as expected
o.x    // fails - shouldn't AnyObject-lookup have found my subscript and called it?
```

I'm not suggesting we make this work, but am suggesting that it could save confusion for users if we just outright ban `@objc` on these subscripts.

Reply:
That restriction wasn't part of the proposal, and I don't see what adding it would achieve: this has nothing to with Objective-C, the same potential for confusion exists if someone did this:

class X {
  subscript(dynamicMember: ... )
}

@dynamicMemberLookup
class Y : X {
}

Because dynamic lookups would be supported if the static type of a reference is Y but not X.

Also, is a subscript like that even @objc compatible?

Reply:
How about the other issue I mentioned, key paths?

I think users might quite reasonably expect to be able to form key paths with this dynamic lookup behavior, but there is no support here for it.

I strongly suspect we'll crash if we build a key path expression with property components for properties that don't actually exist, so it would be great to have a test or two for that.

Reply:
> this has nothing to with Objective-C, the same potential for confusion exists if someone did this

That's a good point. But your example makes me wonder why we would support finding these subscripts in classes that are not either marked `@dynamicmemberlookup` or inheriting from a class that is.

I don't see a test case like:
```
@dynamicmemberlookup class Base {}
class Derived : Base { subscript(dynamicMember: ...) {} }
_ = Derived().dynamicallyLookedUp
```
and I don't see that mentioned in the SE proposal. Is that supposed to work?

Reply:
I just added a testcase for your example:
@dynamicMemberLookup class Base {}

... and the answer is "no" that is not valid. We reject it with this error message:
error: @dynamicMemberLookup attribute requires 'Base' to have a 'subscript(dynamicMember:)' member with a string index